### ISSUE TITLE
Integer querying bug introduced in Realm 3.15.0 and later

### ISSUE THREAD

**Issue by mstralka** [realm/realm-cocoa#6192]
## Goals

I'm pretty sure a bug was introduced in Realm 3.15.0 with how queries are performed when there is an integer comparison in the NSPredicate.

I am able to consistently reproduce this problem in our app, and I think it may have to do with the integer comparison change made in PR #3250, PR #3253, or PR #3272 in 3.15.0-beta.1 because those appear to be related to querying with integer primary keys, which is the problem I'm experiencing. 

This works with Realm 3.13.1
This works with Realm 3.14.1
This works with Realm 3.14.2
It crashes with Realm 3.15.0 (I did not try the betas)
It crashes with Realm 3.16.0
It crashes with Realm 3.16.1
It crashes with Realm 3.16.2

## Expected Results

I have a realm PresentationFile object with ID = 145683 in the local realm file. I can see it in Realm Studio.

With Realm 3.15.0 or later, when I write this query, it does not return the object:
~~~swift
let id = 145683
if let file = presentation.files.filter("id = %@", id).first {
  // 1. update the existing file here - this should be executed because the PresentationFile DOES exist in Realm
} else {
  // 2. create a new PresentationFile here if it's not already in Realm
  // instantiate a new PresentationFile and add to Realm
  realm.add(newFile)
}
~~~

## Actual Results

~~~
Fatal Exception: RLMException
Attempting to create an object of type 'PresentationFile' with an existing primary key value '145683'.
~~~

I can share the full stack trace confidentially.

As a workaround, we reverted to Realm 3.14.2.

## Steps to Reproduce

I can share the test case confidentially in our app.

## Code Sample

I can share confidentially

## Version of Realm and Tooling

Realm framework version: 3.15.0 through at least 3.16.2

Realm Object Server version: N/A

Xcode version: 10.2.1

iOS/OSX version: 11.x, 12.x

Dependency manager + version: Cocoapods 1.6.0

**Comment by nirinchev** [comment]
This appears to be similar to https://github.com/realm/realm-cocoa/issues/6171 - do you think you can email a code sample to support@realm.io as we haven't heard back from the original reporter.

**Comment by bmt2018** [comment]
I was the creator of [#6171](https://github.com/realm/realm-cocoa/issues/6171).  Since I can't submit my actual app, I've been trying to recreate the problem in a separate sample app and until today, I had been unable to reproduce it.  But after seeing this post, it occurred to me that unlike in my production app, my sample app failed to specify the primary key (an integer) on my models.  Once I added that, I was able to reproduce the problem.  I'll submit the sample app to support@realm.io shortly.

**Comment by bmt2018** [comment]
I just submitted my sample app to reproduce this to support@realm.io.

**Comment by ironage** [comment]
Thank you for the reports, please try out the fix in cocoa v3.17.1.

**Comment by gshahbazian** [comment]
@mstralka can you confirm the issue is resolved on version 3.17.1 (https://github.com/realm/realm-cocoa/blob/master/CHANGELOG.md#3171-release-notes-2019-07-10).

We upgraded to 3.17.3 and still see the bug with our filters.
